### DISCUSSION TITLE
Just a mention of an alternative, more generic, strategy for normalised state shape

### DISCUSSION THREAD

**Discussion by Izhaki** [reduxjs/redux#4262] [General]
# Current Docs

The [current docs](https://redux.js.org/usage/structuring-reducers/normalizing-state-shape) give the following example:

```js
    comments : {
        byId : {
            "comment1" : {
                id : "comment1",
                author : "user2",
                comment : ".....",
            },
            "comment2" : {
                id : "comment2",
                author : "user3",
                comment : ".....",
            },
            "comment3" : {
                id : "comment3",
                author : "user3",
                comment : ".....",
            },
            "comment4" : {
                id : "comment4",
                author : "user1",
                comment : ".....",
            },
            "comment5" : {
                id : "comment5",
                author : "user3",
                comment : ".....",
            },
        },
        allIds : ["comment1", "comment2", "comment3", "comment4", "comment5"]
    },
```

And then goes:

> - Any references to individual items should be done by storing the item's ID.
> - Arrays of IDs should be used to indicate ordering.

What might bother some here is the fact you need both `byId` and `allIds`.

# Alternative

Now this is quite a common problem, but for some reason the first thing that popped to my mind when I saw this was [eclipse's GEF framework](https://www.eclipse.org/gef/), where there is a deep nesting requirements (so any `Figure` can have children that are also a `Figure`).

The way they solved it is by adding a (virtual) _root element_, which captures the root level ordering. Like so:

```js
comments : {
    "root-comment" : {
        id : "root-comment",
        children: ["comment1", "comment2"]
    },        
    "comment1" : {
        id : "comment1",
        author : "user2",
        comment : ".....",
    },
    "comment2" : {
        id : "comment2",
        author : "user3",
        comment : ".....",
    },
},
```

This solution may be dubious for flat structures; but for those that do involve deep nesting, the structure above provides a unified way to operate on the data (like adding or removing children).

It is thus a slightly more general solution.

**Comment by markerikson** [comment]
FWIW, the `allIds` pattern was something I based off of Redux-ORM.

With RTK's `createEntityAdapter`, we actually do the same thing, although the field names are `{entities: {}, ids: [] }`, and this allows us to optionally keep the `ids` array in a sorted order.
